presto-main/src/main/java/com/facebook/presto/connector/informationSchema/InformationSchemaPageSourceProvider.java

```diff
@@ -124,17 +129,19 @@ private InternalTable getInternalTable(ConnectorTransactionHandle transactionHan
         InformationSchemaTableHandle handle = split.getTableHandle();
         Map<String, NullableValue> filters = split.getFilters();
 
+        Identity identity = connectorSession.getIdentity();
         Session session = Session.builder(metadata.getSessionPropertyManager())
-                .setTransactionId(transaction.getTransactionId())
                 .setQueryId(new QueryId(connectorSession.getQueryId()))
-                .setIdentity(connectorSession.getIdentity())
+                .setIdentity(new Identity(identity.getUser(), identity.getPrincipal()))
                 .setSource("information_schema")
                 .setCatalog("") // default catalog is not be used
                 .setSchema("") // default schema is not be used
                 .setTimeZoneKey(connectorSession.getTimeZoneKey())
                 .setLocale(connectorSession.getLocale())
                 .setStartTime(connectorSession.getStartTime())
-                .build();
+                .setRole(handle.getCatalogName(), identity.getRole())
+                .build()
+                .beginTransactionId(transaction.getTransactionId(), transactionManager, accessControl);
```

Comment:
I think this entire hack with session recreation needs to be removed at some point. In order to process roles we need to convert session from non transactional to transactional. It won't fail, becuase you cannot query without transaction.